src/main.cpp

```diff
@@ -2451,6 +2451,31 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
         return true;
     }
 
+    // Reject a block that results in a negative shielded value pool balance.
+    if (chainparams.SproutValuePoolCheckpointEnabled()) {
+        // Sprout
+        //
+        // We can expect nChainSproutValue to be valid after the hardcoded
+        // height, and this will be enforced on all descendant blocks. If
+        // the node was reindexed then this will be enforced for all blocks.
+        if (pindex->nChainSproutValue) {
+            if (*pindex->nChainSproutValue < 0) {
+                return state.DoS(100, error("ConnectBlock(): turnstile violation in Sprout shielded value pool"),
+                             REJECT_INVALID, "turnstile-violation-sprout-shielded-pool");
+            }
+        }
+
+        // Sapling
```

Comment:
If we don't gate here, maybe `SproutValuePoolCheckpointEnabled` should be renamed `ValuePoolCheckpointEnabled`?

Reply:
Good point, it may be better referred to as `ZIP209Enabled`.

Reply:
Done.